scripts/data/src/org/akvo/gae/remoteapi/CheckSurveyStructure.java

```diff
@@ -0,0 +1,206 @@
+            }
+        }
+    }
+
+    private void processSurveys(DatastoreService ds) {
+
+        System.out.println("#Processing Surveys");
+
+        final Query survey_q = new Query("Survey");
+        final PreparedQuery survey_pq = ds.prepare(survey_q);
+
+        for (Entity s : survey_pq.asIterable(FetchOptions.Builder.withChunkSize(500))) {
+
+            Long surveyId = s.getKey().getId();
+            String surveyName = (String) s.getProperty("name");
+            surveys.put(surveyId,surveyName);
+        }
+    }
+
+    private void processQuestions(DatastoreService ds) {
+        System.out.println("#Processing Questions");
+
+        final Query qq = new Query("Question");
+        final PreparedQuery qpq = ds.prepare(qq);
+        List<Entity> questionsToFix = new ArrayList<Entity>();
+        List<Key> questionsToKill = new ArrayList<Key>();
+        
+        for (Entity q : qpq.asIterable(FetchOptions.Builder.withChunkSize(500))) {
+
+            Long questionId = q.getKey().getId();
+            Long questionSurvey = (Long) q.getProperty("surveyId");
+            Long questionGroup = (Long) q.getProperty("questionGroupId");
+            String questionText = (String) q.getProperty("text");
+            Long questionGroupSurvey = (Long) qgToSurvey.get(questionGroup);
+
+            if (questionGroup == null || questionGroupSurvey == null) { // in no group or a nonexistent group; hopelessly lost
+                System.out.printf("#ERR: Question %d '%s',survey %d, group %d\n",
+                        questionId, questionText, questionSurvey, questionGroup);
+                orphanQuestions++;
+                if (deleteOrphans){
+                    System.out.println(q.toString());//for posterity
+                    q.setProperty("surveyId", questionGroupSurvey);
+                    questionsToKill.add(q.getKey());
+                }
+            } else { // check for wrong survey/qg
+                qToSurvey.put(questionId, questionSurvey); //ok parent for options
+                if (!questionSurvey.equals(questionGroupSurvey)) {
+                    System.out.printf("#ERR: Question %d '%s' in survey %d, but group %d is in survey %d\n",
+                            questionId, questionText, questionSurvey, questionGroup, questionGroupSurvey);
+                    if (fixSurveyPointers){
+                        System.out.println(q.toString());//for posterity
+                        q.setProperty("surveyId", questionGroupSurvey);
+                        questionsToFix.add(q);
+                    }
+                    unreachableQuestions++;
+                } else {
+                    goodQuestions++;
+                }
+            }
+        }
+        if (questionsToFix.size() > 0) {
+            System.out.printf("#Fixing %d Questions\n",questionsToFix.size());
+            batchSaveEntities(ds, questionsToFix);
+        }
+        if (questionsToKill.size() > 0) {
```

Comment:
also seems more fluid to me to have the `deleteOrphans` as the condition rather than the list size